### PR TITLE
[Frontend] Use symbolic zeros in custom jvp for JAX integration

### LINKED ISSUE #125: [Frontend] Optimize JVP implementation with symbolic zero support
#96 introduced support for JAX transformations such as `grad`, `jacobian`, or `jvp`.

This works by implementing a `custom_jvp` for our QJIT functions. To avoid computing Jacobians for function arguments that do not participate in differentiation, the current implementation checks whether the tangent vector provided for an argument is a concrete value and filled with all zeros.

Recently, JAX has introduced symbolic zeros to be passes as tangent vectors instead of materializing concrete values, which should be more efficient.

Update the implementation of `custom_jvp` to use symbolic zeros once we upgrade our dependency chain to JAX >= 0.4.9.

### LINKED ISSUE #150: [Frontend] JAX integration fails to correctly compute `argnums` in certain cases
Below are two bugs I've encountered using the JAX integration and issues with `custom_jvp` computing the argnums. I'm not sure if these are related or not, so have combined them for now.

Note: they may be addressed by #125.

# Bug 1

Consider

```python
dev = qml.device("lightning.qubit", wires=2)

@qjit
@qml.qnode(dev)
def circuit(x, y):
    qml.RX(x, wires=0)
    qml.RY(y, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(1))

x = jnp.array(0.5)
y = jnp.array(0.1)
```

We see that some combinations of argnums do not work:

```pycon
>>> jax.grad(circuit, argnums=0)(x, y)  # works
Array(-0.47703045, dtype=float64, weak_type=True)
>>> jax.grad(circuit, argnums=1)(x, y)  # fails
File ~/anaconda3/envs/Braket/lib/python3.10/site-packages/catalyst/compilation_pipelines.py:654, in JAX_QJIT.compute_jvp(self, primals, tangents)
    652         deriv_idx = arg_idx * len(results) + res_idx
    653         num_axes = 0 if tangent.ndim == 0 else 1
--> 654         jvp = jnp.tensordot(jnp.transpose(derivatives[deriv_idx]), tangent, axes=num_axes)
    655         jvps[res_idx] = jvps[res_idx] + jvp
    657 if len(results) == 1:

IndexError: list index out of range
```

Using `jax.jit`, it always works:
```pycon
>>> jax.grad(jax.jit(circuit), argnums=0)(x, y)
Array(-0.47703045, dtype=float64, weak_type=True)
>>> jax.grad(jax.jit(circuit), argnums=1)(x, y)
Array(-0.08761211, dtype=float64, weak_type=True)
```

# Bug 2

Consider the following circuit, which has two arguments; one a float and differentiable, and one an integer and not differentiable:

```python
dev = qml.device("lightning.qubit", wires=2)

@qjit
@qml.qnode(dev)
def circuit(params, n):
    
    def ansatz(i, x):
        qml.RX(x[i, 0], wires=0)
        qml.RY(x[i, 1], wires=1)
        qml.CNOT(wires=[0, 1])
        return x

    catalyst.for_loop(0, n, 1)(ansatz)(params.reshape(-1, 2))

    return qml.expval(qml.PauliZ(1))

n = 3
params = jnp.ones(2 * n)
```

Since the second argument is not differentiable, we must pass `argnums=0` when computing the grad:

```pycon
>>> jax.grad(circuit, argnums=0)(params, n)
Array([-0.3262047 ,  0.42661714,  0.16441727, -0.45305815,  0.49515634,
        0.95400696], dtype=float64)
```

However, if `circuit` is jitted with JAX, this no longer works. Internally, `custom_jvp` will determine `argnum=[0, 1]`, which is incorrect:

```pycon
>>> jax.grad(jax.jit(circuit), argnums=0)(params, n)
TypeError: Catalyst.grad only supports differentiation on floating-point arguments, got 'int64' at position 1.
```

This causes issues trying to use Catalyst with jaxopt, as jaxopt will automatically under-the-hood JIT any function that it is trying to optimize.

### PR DESCRIPTION
**Pull request by dime10**
Closes #125 
Closes #150 

[sc-37582]

### CODE REVIEW

File: frontend/catalyst/compilation_pipelines.py [hunk 1]
```diff
@@ -634,10 +634,9 @@ def compute_jvp(self, primals, tangents):
 
         # Optimization: Do not compute Jacobians for arguments which do not participate in
         #               differentiation.
-        # TODO: replace with symbolic zero support
         argnums = []
         for idx, tangent in enumerate(tangents):
-            if qml.math.is_abstract(tangent, like="jax") or jnp.any(jnp.atleast_1d(tangent) != 0.0):
+            if not isinstance(tangent, jax.custom_derivatives.SymbolicZero):
```
Comment (pengmai) [inline frontend/catalyst/compilation_pipelines.py:639]: Will JAX ever pass in a materialized zero tangent (and do we care)? Or is it enough to make a decision based on JAX's behavior?

**Review by pengmai — APPROVED** [review]
Looks great! 🔥

File: frontend/catalyst/compilation_pipelines.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/compilation_pipelines.py:639]: Well with the option for symbolic zeros they shouldn't materialize it if they know the arg doesn't participate in differentiation. Now I think in theory that doesn't prevent the rare case of the tangent vector just happening to be *numerically* zero, but I don't know if we need to consider that.

**Comment by josh146** [comment]
Hey @dime10, does this also resolve the two bugs in #150?

**Comment by dime10** [comment]
Good question, let me give it a try.

**Comment by josh146** [comment]
Awesome :) If it does fix them, I suggest adding those two minimal examples as tests in this PR, to ensure that behaviour is tested from hereon

**Comment by dime10** [comment]
@josh146 Bug 2 was indeed fixed by the symbolic zeros, Bug 1 was not but was an easy fix in the custom_jvp function.

File: frontend/test/pytest/test_jax_integration.py [hunk 2]
```diff
@@ -325,6 +326,27 @@ def cost_fn(x, y):
         assert "0" in circuit.jaxed_qfunc.deriv_qfuncs
         assert len(circuit.jaxed_qfunc.deriv_qfuncs["0"].jaxpr.out_avals) == 1
 
+    def test_jit_and_grad(self, backend):
+        """Test that argnum determination works correctly when combining jax.jit with jax.grad.
+        This was fixed by the introduction of symbolic zero detection for tangent vectors."""
```
Comment (josh146) [inline frontend/test/pytest/test_jax_integration.py:331]: @dime10 a suggestion: to use the `monkeypatch` fixture here to ensure that the internal argnums _are_ being set correctly :)

**Comment by josh146** [comment]
@dime10 awesome 💯 Does the test you added cover both bugs?

**Comment by dime10** [comment]
> @dime10 awesome 💯 Does the test you added cover both bugs?

Yes I added two tests (one new function and one modified via parametrization of the inputs).

**Review by erick-xanadu — APPROVED** [review]
Looks great! :100:

File: frontend/test/pytest/test_jax_integration.py (hunk 2 above)
Comment (pengmai) [reply frontend/test/pytest/test_jax_integration.py:332]: This has been implemented in the `test_argnums_passed` test in this file.

**Review by josh146 — APPROVED** [review]
amazing, thanks @pengmai!

**Pull request merged**